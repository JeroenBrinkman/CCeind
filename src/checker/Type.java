package checker;

import iloc.eval.Machine;

/** TempName data type. */
abstract public class Type {
	/** The singleton instance of the {@link Bool} type. */
	public static final Type BOOL = new Bool();
	/** The singleton instance of the {@link Int} type. */
	public static final Type INT = new Int();
	/** The singleton instance of the {@link Char} type. */
	public static final Type CHAR = new Char();
	/** The singleton instance of the {@link String} type. */
	public static final Type STRING = new String();
	/** The singleton instance of the {@link Void} type. */
	public static final Type VOID = new Void();
	
	private final TypeKind kind;

	/** Constructor for subclasses. */
	protected Type(TypeKind kind) {
		this.kind = kind;
	}

	/** Returns the kind of this type. */
	public TypeKind getKind() {
		return this.kind;
	}

	/** returns the size (in bytes) of a value of this type. */
	abstract public int size();

	/** Representation of the Boolean type. */
	static public class Bool extends Type {
		private Bool() {
			super(TypeKind.BOOL);
		}

		@Override
		public int size() {
			return Machine.INT_SIZE;
		}

		@Override
		public java.lang.String toString() {
			return "Boolean";
		}
	}

	/** Representation of the Integer type. */
	static public class Int extends Type {
		private Int() {
			super(TypeKind.INT);
		}

		@Override
		public int size() {
			return Machine.INT_SIZE;
		}

		@Override
		public java.lang.String toString() {
			return "Integer";
		}
	}
	
	/** Representation of the Char type. */
	static public class Char extends Type{

		private Char(){
			super(TypeKind.CHAR);
		}
		
		@Override
		public int size() {
			return Machine.DEFAULT_CHAR_SIZE;
		}

		@Override
		public java.lang.String toString() {
			return "Character";
		}
		
	}
	
	/** Representation of the String type. */
	static public class String extends Type{
		private int length = 0;
		private String(){
			super(TypeKind.STRING);
		}
		
		@Override
		public int size() {
			return length*Machine.DEFAULT_CHAR_SIZE;
		}

		@Override
		public java.lang.String toString() {
			return "String";
		}
		
		public void setLength(int length){
			this.length = length;
			
		}
		
	}
	
	/** Representation of the Void type. */
	static public class Void extends Type{

		private Void(){
			super(TypeKind.VOID);
		}
		
		@Override
		public int size() {
			return 0;
		}

		@Override
		public java.lang.String toString() {
			return "Void";
		}
		
	}
}
