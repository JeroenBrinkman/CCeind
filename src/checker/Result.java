package checker;

import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.ParseTreeProperty;

/** Class holding the results of the TempName checker. */
public class Result {
	/**
	 * Mapping from statements and expressions to the atomic subtree that is
	 * their entry in the control flow graph.
	 */
	private final ParseTreeProperty<ParserRuleContext> entries = new ParseTreeProperty<>();
	/** Mapping from expressions to types. */
	private final ParseTreeProperty<Type> types = new ParseTreeProperty<>();
	private final ParseTreeProperty<Type[]> readtypes = new ParseTreeProperty<>();

	/** Adds an association from parse tree node to the flow graph entry. */
	public void setEntry(ParseTree node, ParserRuleContext entry) {
		this.entries.put(node, entry);
	}

	/**
	 * Returns the flow graph entry associated with a given parse tree node.
	 */
	public ParserRuleContext getEntry(ParseTree node) {
		return this.entries.get(node);
	}

	/**
	 * Adds an association from a parse tree expression, type, or assignment
	 * target node to the corresponding (inferred) type.
	 */
	public void setType(ParseTree node, Type type) {
		this.types.put(node, type);
	}

	/** Returns the type associated with a given parse tree node. */
	public Type getType(ParseTree node) {
		return this.types.get(node);
	}

	public void setReadTypes(ParseTree node, Type[] tps) {
		this.readtypes.put(node, tps);
	}

	public Type[] getReadTypes(ParseTree node) {
		return this.readtypes.get(node);
	}
}
