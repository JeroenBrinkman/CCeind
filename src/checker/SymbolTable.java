package checker;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class SymbolTable{

	private final Deque<Map<String, Type>> stack = new ArrayDeque<Map<String,Type>>();
	private List<String> constants = new ArrayList<String>();
	
	public void openScope() {
		stack.push(new HashMap<String, Type>());
	}

	public void closeScope() {
		stack.pop();
	}

	public boolean add(String id, Type type) {
		if(contains(id) || isConst(id)){
			return false;
			
		}else{
			Map<String, Type> currentScope;
			currentScope = stack.getFirst();
			currentScope.put(id, type);
			return true;
		}
	}
	
	public void addConst(String id){
		constants.add(id);
	}

	public boolean contains(String id) {
		Map<String, Type> currentScope;
		for(Iterator<Map<String, Type>> itr = stack.iterator(); itr.hasNext();){
			currentScope = (Map<String, Type>) itr.next();
			if(currentScope.containsKey(id)){
				return true;
			}
		}
		return false;
	}
	
	public boolean isConst(String id){
		return constants.contains(id);
	}
	
	public Type getType(String id){
		Type type = Type.VOID;
		Map<String, Type> currentScope;
		for(Iterator<Map<String, Type>> itr = stack.descendingIterator(); itr.hasNext();){
			currentScope = (Map<String, Type>) itr.next();
			if(currentScope.containsKey(id)){
				type = currentScope.get(id);
			}
		}
		return type;
	}
	
}
