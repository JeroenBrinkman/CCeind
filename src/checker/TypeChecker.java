package checker;

import grammar.TempNameBaseListener;
import grammar.TempNameParser;
import grammar.TempNameParser.ExprContext;
import grammar.TempNameParser.*;

import java.util.ArrayList;
import java.util.List;

import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.misc.NotNull;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.ParseTreeWalker;

/** Class to type check and calculate flow entries and variable offsets. */
public class TypeChecker extends TempNameBaseListener {
	/** Result of the latest call of {@link #check}. */
	private Result result;
	/** List of errors collected in the latest call of {@link #check}. */
	private List<String> errors;
	/** Variable scope for the latest call of {@link #check}. */
	private SymbolTable sT;

	/**
	 * Runs this checker on a given parse tree, and returns the checker result.
	 * 
	 * @throws ParseException
	 *             if an error was found during checking.
	 */
	public Result check(ParseTree tree) throws ParseException {
		this.result = new Result();
		this.errors = new ArrayList<>();
		new ParseTreeWalker().walk(this, tree);
		if (hasErrors()) {
			throw new ParseException(getErrors());
		}
		return this.result;
	}

	/** Indicates if any errors were encountered in this tree listener. */
	public boolean hasErrors() {
		return !getErrors().isEmpty();
	}

	/** Returns the list of errors collected in this tree listener. */
	public List<String> getErrors() {
		return this.errors;
	}

	/**
	 * Checks the inferred type of a given parse tree, and adds an error if it
	 * does not correspond to the expected type.
	 */
	private void checkType(ParserRuleContext node, Type expected) {
		Type actual = getType(node);
		if (actual == null) {
			throw new IllegalArgumentException("Missing inferred type of " + node.getText());
		}
		if (!actual.equals(expected)) {
			addError(node, "Expected type '%s' but found '%s'", expected, actual);
		}
	}

	/**
	 * Checks the inferred type of a given parse tree, and adds an error if it
	 * does correspond to the expected type.
	 */
	private void checkNotType(ParserRuleContext node, Type expected) {
		Type actual = getType(node);
		if (actual == null) {
			throw new IllegalArgumentException("Missing inferred type of " + node.getText());
		}
		if (actual.equals(expected)) {
			addError(node, "Illegal type '%s' not allowed", expected);
		}
	}

	/**
	 * Records an error at a given parse tree node.
	 * 
	 * @param ctx
	 *            the parse tree node at which the error occurred
	 * @param message
	 *            the error message
	 * @param args
	 *            arguments for the message, see {@link String#format}
	 */
	public void addError(ParserRuleContext node, String message, Object... args) {
		addError(node.getStart(), message, args);
	}

	/**
	 * Records an error at a given token.
	 * 
	 * @param token
	 *            the token at which the error occurred
	 * @param message
	 *            the error message
	 * @param args
	 *            arguments for the message, see {@link String#format}
	 */
	private void addError(Token token, String message, Object... args) {
		int line = token.getLine();
		int column = token.getCharPositionInLine();
		message = String.format(message, args);
		message = String.format("Line %d:%d - %s", line, column, message);
		this.errors.add(message);
	}

	/** Convenience method to add a type to the result. */
	private void setType(ParseTree node, Type type) {
		this.result.setType(node, type);
	}

	/** Returns the type of a given expression or type node. */
	private Type getType(ParseTree node) {
		return this.result.getType(node);
	}

	/** Convenience method to add a flow graph entry to the result. */
	private void setEntry(ParseTree node, ParserRuleContext entry) {
		if (entry == null) {
			throw new IllegalArgumentException("Null flow graph entry");
		}
		this.result.setEntry(node, entry);
	}

	/** Returns the flow graph entry of a given expression or statement. */
	private ParserRuleContext entry(ParseTree node) {
		return this.result.getEntry(node);
	}

	@Override
	public void enterProgram(@NotNull TempNameParser.ProgramContext ctx) {
		sT = new SymbolTable();
		sT.openScope();
	}

	@Override
	public void exitProgram(@NotNull TempNameParser.ProgramContext ctx) {
		sT.closeScope();
	}

	@Override
	public void enterBlockExpr(BlockExprContext ctx) {
		sT.openScope();
	}

	@Override
	public void exitTrueExpr(TrueExprContext ctx) {
		setType(ctx, Type.BOOL);
		setEntry(ctx, ctx);
	}

	@Override
	public void exitBoolType(BoolTypeContext ctx) {
		setType(ctx, Type.BOOL);
	}

	@Override
	public void exitStringExpr(StringExprContext ctx) {
		setType(ctx, Type.STRING);
		setEntry(ctx, ctx);
	}

	@Override
	public void exitParExpr(ParExprContext ctx) {
		setType(ctx, getType(ctx.expr()));
		setEntry(ctx, entry(ctx.expr()));
	}

	@Override
	public void exitCompExpr(CompExprContext ctx) {
		if (ctx.compOp().getText().equals("<>") || ctx.compOp().getText().equals("==")) {
			checkType(ctx.expr(1), getType(ctx.expr(0)));
		} else {
			checkType(ctx.expr(0), Type.INT);
			checkType(ctx.expr(1), Type.INT);
		}
		setType(ctx, Type.BOOL);
		setEntry(ctx, entry(ctx.expr(0)));
	}

	@Override
	public void exitStringType(StringTypeContext ctx) {
		setType(ctx, Type.STRING);
	}

	@Override
	public void exitIfExpr(IfExprContext ctx) {
		checkType(ctx.expr(0), Type.BOOL);
		setEntry(ctx, entry(ctx.expr(0)));
		if (ctx.expr(2) != null) {
			checkType(ctx.expr(1), getType(ctx.expr(2)));
		}
		setType(ctx, getType(ctx.expr(1)));
	}

	@Override
	public void exitBlockExpr(BlockExprContext ctx) {
		ExprContext last = ctx.expr(ctx.expr().size() - 1);
		setType(ctx, getType(last));
		if (ctx.expr().size() > 0) {
			setEntry(ctx, entry(ctx.expr(0)));
		}
		sT.closeScope();
	}

	@Override
	public void exitFalseExpr(FalseExprContext ctx) {
		setType(ctx, Type.BOOL);
		setEntry(ctx, ctx);
	}

	@Override
	public void exitPrintExpr(PrintExprContext ctx) {
		for (ExprContext e : ctx.expr()) {
			checkNotType(e, Type.VOID);
		}
		if (ctx.expr().size() == 1) {
			setType(ctx, getType(ctx.expr(0)));
			setEntry(ctx, entry(ctx.expr(0)));
		} else {
			setEntry(ctx, entry(ctx.expr(0)));
			setType(ctx, Type.VOID);
		}
	}

	@Override
	public void exitCharType(CharTypeContext ctx) {
		setType(ctx, Type.CHAR);
	}

	@Override
	public void exitIdTarget(IdTargetContext ctx) {
		final String id = ctx.ID().getText();
		if (!sT.contains(id)) {
			addError(ctx, "Variable '%s' not declared", id);
			setEntry(ctx, ctx);
			setType(ctx, Type.VOID);
		} else {
			setType(ctx, sT.getType(id));
			setEntry(ctx, ctx);
			// setOffset(ctx, sT.offset(id));
		}
	}

	@Override
	public void exitCharExpr(CharExprContext ctx) {
		setType(ctx, Type.CHAR);
		setEntry(ctx, ctx);
	}

	@Override
	public void exitIntType(IntTypeContext ctx) {
		setType(ctx, Type.INT);
	}

	@Override
	public void exitReadExpr(ReadExprContext ctx) {
		Type[] tps = new Type[ctx.ID().size()];
		for (int i = 0; i < ctx.ID().size(); i++) {
			String targetString = ctx.ID(i).toString();
			if (!sT.contains(targetString)) {
				errors.add(" Attempting to read for a undeclared variable '" + ctx.ID().toString() + "' !");
			}
			tps[i] = sT.getType(ctx.ID(i).getText());
		}

		result.setReadTypes(ctx, tps);
		if (ctx.ID().size() > 1) {
			setType(ctx, Type.VOID);
			setEntry(ctx, ctx);
		} else {
			setType(ctx, sT.getType(ctx.ID(0).getText()));
			setEntry(ctx, ctx);
		}

	}

	@Override
	public void exitMultExpr(MultExprContext ctx) {
		checkType(ctx.expr(0), Type.INT);
		checkType(ctx.expr(1), Type.INT);
		setType(ctx, Type.INT);
		setEntry(ctx, entry(ctx.expr(0)));
	}

	@Override
	public void exitNumExpr(NumExprContext ctx) {
		setType(ctx, Type.INT);
		setEntry(ctx, ctx);
	}

	@Override
	public void exitPlusExpr(PlusExprContext ctx) {
		if (getType(ctx.expr(0)).equals(Type.STRING)) {
			if (getType(ctx.expr(0)).equals(Type.STRING) || getType(ctx.expr(0)).equals(Type.CHAR)) {
				setType(ctx, Type.STRING);
			}else{
				addError(ctx, "Illegal type second argument, STRING or CHAR expected.");
				setType(ctx, Type.VOID);
			}
		} else if (getType(ctx.expr(0)).equals(Type.INT)) {
			checkType(ctx.expr(1), Type.INT);
			setType(ctx, Type.INT);
		} else {
			addError(ctx, "Illegal type first argument, STRING or INT expected.");
			setType(ctx, Type.VOID);
		}
		setEntry(ctx, entry(ctx.expr(0)));
	}

	@Override
	public void exitDeclExpr(DeclExprContext ctx) {
		if (ctx.expr() != null) {
			checkType(ctx.expr(), getType(ctx.type()));
			setEntry(ctx, entry(ctx.expr()));
			if (!sT.add(ctx.ID().getText(), getType(ctx.type()))) {
				addError(ctx, "Illegal declaration, '$s' already in scope.", ctx.ID().getText());
			} else {

				setType(ctx.ID(), getType(ctx.type()));
				setType(ctx, getType(ctx.type()));
				return;
			}
		} else {
			if (!sT.add(ctx.ID().getText(), getType(ctx.type()))) {
				addError(ctx, "Illegal declaration, '$s' already in scope.", ctx.ID().getText());

			}
			// TODO entry?
		}
		setType(ctx, Type.VOID);
	}

	@Override
	public void exitWhileExpr(WhileExprContext ctx) {
		checkType(ctx.expr(0), Type.BOOL);
		setEntry(ctx, entry(ctx.expr(0)));
		setType(ctx, Type.VOID);
	}

	@Override
	public void exitAssExpr(AssExprContext ctx) {
		checkType(ctx.expr(), getType(ctx.target()));
		setType(ctx, getType(ctx.target()));
		setEntry(ctx, entry(ctx.expr()));
	}

	@Override
	public void exitPrfExpr(PrfExprContext ctx) {
		Type type;
		if (ctx.prfOp().MINUS() != null) {
			type = Type.INT;
		} else {
			type = Type.BOOL;
		}
		checkType(ctx.expr(), type);
		setType(ctx, type);
		setEntry(ctx, entry(ctx.expr()));
	}

	@Override
	public void exitBoolExpr(BoolExprContext ctx) {
		checkType(ctx.expr(0), Type.BOOL);
		checkType(ctx.expr(1), Type.BOOL);
		setType(ctx, Type.BOOL);
		setEntry(ctx, entry(ctx.expr(0)));
	}

	@Override
	public void exitIdExpr(IdExprContext ctx) {
		String id = ctx.ID().getText();
		Type type = this.sT.getType(id);
		if (type == null) {
			addError(ctx, "Variable '%s' not declared", id);
			setType(ctx, Type.VOID);
		} else {
			setType(ctx, type);
			// setOffset(ctx, this.scope.offset(id));
			setEntry(ctx, ctx);
		}
	}

}