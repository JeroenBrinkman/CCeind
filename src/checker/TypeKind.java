package checker;

/** TempName data type kind. */
public enum TypeKind {
	/** Integer base type. */
	INT,
	/** Boolean base type. */
	BOOL,
	/** Character base Type*/
	CHAR,
	/** String base Type*/
	STRING,
	/** Void base Type*/
	VOID;
}
