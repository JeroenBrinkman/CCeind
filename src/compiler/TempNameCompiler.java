package compiler;


import java.io.File;
import java.io.FileReader;
import java.io.IOException;

import org.antlr.v4.runtime.ANTLRInputStream;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.Lexer;
import org.antlr.v4.runtime.TokenStream;
import org.antlr.v4.runtime.tree.ParseTree;

import checker.ErrorListener;
import checker.ParseException;
import checker.Result;
import checker.TypeChecker;
import grammar.TempNameLexer;
import grammar.TempNameParser;

public class TempNameCompiler {
	/** The singleton instance of this class. */
	private final static TempNameCompiler instance = new TempNameCompiler();
	/** Debug flag. */
	private final static boolean SHOW = true;

	/** Returns the singleton instance of this class. */
	public static TempNameCompiler instance() {
		return instance;
	}

	/** Compiles and runs the program named in the argument. */
	public static void main(String[] args) {
		if (args.length != 1) {
			System.err.println("Usage: filename");
			return;
		}
		//try {
			System.out.println("--- Running " + args[0]);
			//Program prog = instance().compile(new File(args[0]));
			if (SHOW) {
				//System.out.println(prog.prettyPrint());
			}
			//Simulator sim = new Simulator(prog);
			//sim.run();
			System.out.println("--- Done with " + args[0]);
		/**} catch (ParseException exc) {
			exc.print();
		} catch (IOException exc) {
			exc.printStackTrace();
		}*/
	}

	/** The fixed checker of this compiler. */
	private final TypeChecker checker;
	/** The fixed generator of this compiler. */
	//private final Generator generator;

	private TempNameCompiler() {
		this.checker = new TypeChecker();
		//this.generator = new Generator();
	}

	/** Typechecks a given Simple Pascal string. */
	public Result check(String text) throws ParseException {
		return check(parse(text));
	}

	/** Typechecks a given Simple Pascal file. */
	public Result check(File file) throws ParseException, IOException {
		return check(parse(file));
	}

	/** Typechecks a given Simple Pascal parse tree. */
	public Result check(ParseTree tree) throws ParseException {
		return this.checker.check(tree);
	}

	/** Compiles a given Simple Pascal string into an ILOC program. 
	public Program compile(String text) throws ParseException {
		return compile(parse(text));
	}*/

	/** Compiles a given Simple Pascal file into an ILOC program. 
	public Program compile(File file) throws ParseException, IOException {
		return compile(parse(file));
	}*/

	/** Compiles a given Simple Pascal parse tree into an ILOC program. 
	public Program compile(ParseTree tree) throws ParseException {
		Result checkResult = this.checker.check(tree);
		
		return this.generator.generate(tree, checkResult);
	}*/

	/** Compiles a given Simple Pascal string into a parse tree. */
	public ParseTree parse(String text) throws ParseException {
		return parse(new ANTLRInputStream(text));
	}

	/** Compiles a given Simple Pascal string into a parse tree. */
	public ParseTree parse(File file) throws ParseException, IOException {
		return parse(new ANTLRInputStream(new FileReader(file)));
	}

	private ParseTree parse(CharStream chars) throws ParseException {
		ErrorListener listener = new ErrorListener();
		Lexer lexer = new TempNameLexer(chars);
		lexer.removeErrorListeners();
		lexer.addErrorListener(listener);
		TokenStream tokens = new CommonTokenStream(lexer);
		TempNameParser parser = new TempNameParser(tokens);
		parser.removeErrorListeners();
		parser.addErrorListener(listener);
		ParseTree result = parser.program();
		System.out.println(result.toString());
		listener.throwException();
		return result;
	}
}
