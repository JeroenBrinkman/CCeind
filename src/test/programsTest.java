package test;

import static org.junit.Assert.*;
import iloc.Simulator;
import iloc.model.Program;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;

import org.junit.Test;

import checker.ParseException;
import compiler.TempNameCompiler;

public class programsTest {
	private final static String BASE_DIR = "src/files";
	private final static String EXT = ".tmp";
	private final TempNameCompiler compiler = TempNameCompiler.instance();

	@Test
	public void gcd() {
		System.out.println("----Testing GCD----");
		Program prog;
		try {
			prog = compile("gcd");
			System.out.println(prog.prettyPrint());
			String out = sim(prog, "100\n170");
			System.out.println(out);
			
		} catch (IOException e) {
			fail("Read/Write fault");
		} catch (ParseException e) {
			System.err.println("Errors: ");
			for (String msg : e.getMessages()) {
				System.err.println(msg);
			}
			fail("An error occured, see command line for more information.");
		}
	}
	
	@Test
	public void fib() {
		System.out.println("----Testing Fib----");
		Program prog;
		try {
			prog = compile("fib");
			System.out.println(prog.prettyPrint());
			String out = sim(prog, "8");
			System.out.println(out);
			
		} catch (IOException e) {
			fail("Read/Write fault");
		} catch (ParseException e) {
			System.err.println("Errors: ");
			for (String msg : e.getMessages()) {
				System.err.println(msg);
			}
			fail("An error occured, see command line for more information.");
		}
	}

	private Program compile(String filename) throws IOException, ParseException {
		return this.compiler.compile(new File(BASE_DIR, filename + EXT));
	}

	private String sim(Program prog, String input) {
		Simulator sim = new Simulator(prog);
		sim.setIn(new ByteArrayInputStream(input.getBytes()));
		ByteArrayOutputStream out = new ByteArrayOutputStream();
		sim.setOut(out);
		sim.run();
		return out.toString();
	}

}
