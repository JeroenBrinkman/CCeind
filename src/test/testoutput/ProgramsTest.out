----Testing Fib----
Program:   nop                                   
           in      "n0? : "  => r_0              
           storeAI r_0       => r_arp,0          
           loadAI  r_arp,0   => r_0              
           loadI   1         => r_1              
           loadAI  r_arp,0   => r_2              
           cmp_LE  r_2,r_1   => r_3              
           cbr     r_3       -> n_7_8,n_9_8      
n_7_8:     nop                                   
           loadI   1         => r_4              
           out     "1: ",r_4                     
           storeAI r_4       => r_arp,4          
           loadAI  r_arp,4   => r_5              
           storeAI r_5       => r_arp,4          
           storeAI r_5       => r_arp,8          
           jumpI             -> endIf_6_1        
n_9_8:     nop                                   
           loadI   1         => r_6              
           storeAI r_6       => r_arp,12         
           loadI   1         => r_7              
           storeAI r_7       => r_arp,16         
           loadI   1         => r_8              
           storeAI r_8       => r_arp,20         
n_12_8:    nop                                   
           loadAI  r_arp,20  => r_9              
           loadAI  r_arp,0   => r_10             
           cmp_LT  r_9,r_10  => r_11             
           cbr     r_11      -> n_13_19,end_12_2 
n_13_19:   nop                                   
           loadAI  r_arp,12  => r_12             
           loadAI  r_arp,16  => r_13             
           add     r_12,r_13 => r_14             
           storeAI r_14      => r_arp,24         
           loadAI  r_arp,12  => r_15             
           storeAI r_15      => r_arp,16         
           loadAI  r_arp,24  => r_16             
           storeAI r_16      => r_arp,12         
           loadI   1         => r_17             
           loadAI  r_arp,20  => r_18             
           add     r_18,r_17 => r_19             
           storeAI r_19      => r_arp,20         
           loadAI  r_arp,20  => r_20             
           jumpI             -> n_12_8           
end_12_2:  nop                                   
           loadAI  r_arp,12  => r_21             
           out     "n1: ",r_21                   
           storeAI r_21      => r_arp,28         
           loadAI  r_arp,28  => r_22             
           storeAI r_22      => r_arp,24         
           storeAI r_22      => r_arp,8          
endIf_6_1: nop                                   
           loadAI  r_arp,8   => r_23             
           storeAI r_23      => r_arp,12         

n1: 34

----Testing GCD----
Program: nop                               
         in      "a? : "  => r_0           
         storeAI r_0      => r_arp,0       
         loadAI  r_arp,0  => r_0           
         in      "b? : "  => r_1           
         storeAI r_1      => r_arp,4       
         loadAI  r_arp,4  => r_1           
n_7_7:   nop                               
         loadI   0        => r_2           
         loadAI  r_arp,4  => r_3           
         cmp_NE  r_3,r_2  => r_4           
         cbr     r_4      -> n_8_7,end_7_1 
n_8_7:   nop                               
         loadAI  r_arp,4  => r_5           
         storeAI r_5      => r_arp,8       
         loadAI  r_arp,0  => r_6           
         loadAI  r_arp,4  => r_7           
         div     r_6,r_7  => r_8           
         mult    r_8,r_7  => r_8           
         sub     r_6,r_8  => r_9           
         storeAI r_9      => r_arp,4       
         loadAI  r_arp,8  => r_10          
         storeAI r_10     => r_arp,0       
         loadAI  r_arp,0  => r_11          
         jumpI            -> n_7_7         
end_7_1: nop                               
         loadAI  r_arp,0  => r_12          
         out     "a: ",r_12                
         storeAI r_12     => r_arp,12      
         loadAI  r_arp,12 => r_13          
         storeAI r_13     => r_arp,8       

a: 10